man/lformat.man
.\" 3.2
.\"  /*******************************************************************\
.\"   * Copyright (C) 1995 Lars J Aas <larsa@colargol.stud.idb.hist.no> *
.\"   * based on documentation by C J Needham <col@imdb.com> 1995,      *
.\"   * permission is granted by the authors to freely distribute       *
.\"   *                        providing no fee of any kind is charged. *
.\"  \*******************************************************************/
.\"
.TH LFORMAT 1 "RELEASEDATE" " " "The Internet Movie Database vVERSIONNUMBER"
.SH NAME
lformat \- a program for formatting lists 
.SH SYNOPSYS
.B
lformat [options]
.SH DESCRIPTION
This program provides a number of features for formatting whole databases,
combining the information with the movies list or the movie ratings report. 
.SH DATABASE OPTIONS
These options specify which database you would like to format:
.TP
.B \-acr
actors database
.TP
.B \-acs
actresses database
.TP
.B \-dir
directors database
.TP
.B \-write
writers database
.TP
.B \-comp
composers database
.TP
.B \-cine
cinematographers database
.TP
.B \-edit
editors database
.TP
.B \-prodes
production designers database
.TP
.B \-costdes
costume designers database
.TP
.B \-prdcr
producers database
.TP
.B \-misc
miscellaneous database
.SH FORMAT OPTIONS
.TP
.B \-yr
Adds the year of release information and sorts each person's entries
into chronological order.
.TP
.B \-yru
Adds the year of release information, but leaves entries in
alphabetical order.
.TP
.B \-mrr
Adds the information from the movie ratings report to each title.
.TP
.B \-smrr
Adds the information from the movie ratings report to each title and
sorts each person's entries by average rating.
.TP
.B \-vmrr
Adds the information from the movie ratings report to each title and
sorts each person's entries by number of votes.
.PP
If no format option is given, the program formats the results in the style of
the published lists.
.SH OUTPUT OPTIONS
.TP
.B \-r
Raw mode, don't format results.
.TP
.B \-m
Movies only, ignores TV-series.
.SH EXAMPLES
.TP
Regenerate published form of actors list:
.B
lformat -acr
.TP
Format directors database and add year information:
.B
lformat -dir -yr
.TP
Format actress list and add ratings report information:
.B
lformat -acs -mrr


docs/ChangeLog
-------Version 3.4 Released---------------------

Sun Mar 29 22:40:37 BST 1998

 * biographies list offset fields increased to 4 bytes following overflow
 * various constants updated

-------Version 3.3c Released---------------------

Sat Apr 26 06:51:19 BST 1997

 * added support for crew completion list
 * added support for distributors list
 * adapted mkdbs for more robust processing of filmography lists
 * added support for new business list tag

-------Version 3.3b Released---------------------

Sat Mar 15 08:09:50 GMT 1997

 * added support for ISO-Latin-1

-------Version 3.3 Released---------------------

Wed Oct 16 22:28:54 BST 1996

 * added support for AKA attributes
 * added support for writers ordering
 * moved misc section to after case in title output
 * added support for business list
 * added support for laserdisc list
 * fixed tagline bug
 * diplays URL of web page for name/title

-------Version 3.2h Released---------------------

Tue Oct  8 14:24:21 BST 1996

 * added support for special effects companies list
 * updated various system constants
 * added new movie link types 'edited into' & 'edited from'
 * added extra checks to avoid title/attribute key file corruption in mkdbs.c

-------Version 3.2f Released---------------------

Mon Jun 24 22:04:46 GMT 1996

 * added support for language list

-------Version 3.2e Released---------------------

Sat Mar 16 08:09:39 GMT 1996

 * added support for alternate versions list
 * added bio keys IT & AT

-------Version 3.2d Released---------------------

Tue Aug 15 23:23:47 GMT 1995

 * switched TitleID to unsigned long to handle >65K titles
 * fixed bug in display of literature and genre data
 * added bio key WN

-------Version 3.2c Released---------------------

Sun May 14 23:00:14 GMT 1995

 * raised MAXFILMOGRAPHIES constant to handle larger no. of entries

-------Version 3.2 Released---------------------

Sun Jan 29 22:40:08 GMT 1995

 * added full support for title searching without year/TV/V info

Tue Jan  3 23:00:45 GMT 1995

 * replaced 'biodefs' file
 * fixed bug when looking for common titles with people using -name

Mon Jan  2 21:52:08 GMT 1995

 * log file support
 * better separation of search/display code

Wed Dec 14 21:43:57 GMT 1994

 * fixed bug in 'title' for miscellaneous section
 * added support for the movie links list

Tue Dec 13 22:06:44 GMT 1994
 
 * noted when cast in credits output
 * fixed name ordering problem for names with (I)/(II) marker
 * display Actor/Actress in list -name where appropriate

Wed Dec  7 21:19:19 GMT 1994

 * added support for tag lines & complete cast list
 * put spaces around "." in cast results
 * standardized number of dashes printed out in 'title' / 'list'

Mon Dec  5 21:19:42 GMT 1994

 * multi-country productions displayed on same line

Wed Nov 30 18:14:16 GMT 1994

 * switched to new IMDB license
 * updated 'lfetch' to get new lists
 * fixed "No plot summaries message"
 * simplified MRR processing in mkdb
 * added support for locations & release date lists
 * switched 'Reviews' marker for clarity
 * added support for technical & literature lists

Mon Oct 31 21:42:03 GMT 1994
 
-------Version 3.1 Released---------------------

Mon Oct 17 19:04:26 GMT 1994
 
 * 3.1 code frozen
 * minor bug fixes with lindex

Sun Oct  9 18:37:14 GMT 1994

 * added support for countries list

Tue Oct  4 21:27:53 GMT 1994

 * added support for cast ordering info

Sun Oct  2 23:02:59 GMT 1994

 * added support for color info and sound mix lists
 * added constraint options to 'list'
 * standardized list options

Wed Sep 28 22:22:56 GMT 1994

 * re-ordered title info printing plus separated genres list
 * print genres on one line

Mon Sep 26 21:53:47 GMT 1994

 * added support for production companies list

Sun Sep 25 22:02:07 GMT 1994

 * added constraint options to 'lindex'

Tue Sep  6 20:09:52 GMT 1994

 * added support for certificates & genres list
 * fixed mkballot bug with -ar option

Tue Aug 23 20:08:04 GMT 1994

 * added support for miscellaneous filmographies list

Tue May 24 19:13:47 GMT 1994

 * moviedb-3.0 completed

Thu May 19 22:35:12 GMT 1994

 * switched to merged character names scheme

Tue May 17 21:09:45 GMT 1994

 * new 'mkdb' eliminating need for large array to hold name keys
 * removed duplicate aka titles in 'title -s -aka' queries

Wed May 11 22:11:59 GMT 1994

 * made upperCase a macro for increased speed
 * switched to Phil's new Makefiles

Wed May  4 17:46:08 GMT 1994

 * mapped case of titles/names to match key files
 * added support for running times list
 * lguide error messages

Tue May  3 21:38:30 GMT 1994
 
 * added support for Phil's imoviedb package
 * expanded compression support

Thu Apr 28 19:51:52 GMT 1994

 * finished rewrite to use new database representation

-------Version 2.9 Released---------------------

Wed Mar 16 21:39:40 GMT 1994

 * added support for producers list
 * added support for soundtracks list
 * added support for generic aka lists

Tue Mar 15 23:47:46 GMT 1994

 * replaced -crazy and -triv options with -trivia (also includes goofs & quotes)
 * added support for quotes list
 * added support for goofs list
 * handled -full error messages correctly
 * correct conditions for printing of missing plot summaries message
 * fixed lindex to search all lists when no list options given
 * -mrr no longer implies -m

Mon Mar 14 21:48:21 GMT 1994

 * new version of check-for-program
 * passed $MAKE variable to make subprocesses
 * lfetch modified to match full filenames
 * added 'keepchars' option for CLEANLISTS

-------Version 2.8 Released---------------------

Wed Dec  1 21:58:21 GMT 1993

 * man pages added to package
 * LaTeX documentation added to package

Sat Nov 27 20:53:40 GMT 1993

 * 2.8 code finished!

Thu Nov 25 18:17:00 GMT 1993

 * ported lseen to C

Tue Nov 23 19:16:48 GMT 1993

 * added gendb -votes option

Mon Nov 22 23:54:58 GMT 1993

 * ported lformat to C
 * removed yrf.c file

Wed Nov 17 22:36:44 GMT 1993


src/lformat.c
/*============================================================================
 *
 *  Program: lformat.c 
 *
 *  Version: 3.18
 *
 *  Purpose: format list databases
 *
 *  Author:  C J Needham <col@imdb.com>     
 *
 *  Copyright (c) 1996-2002 The Internet Movie Database Inc.
 *
 *  Permission is granted by the copyright holder to distribute this program
 *  is source form only, providing this notice remains intact, and no fee
 *  of any kind is charged. This software is provided as is and there are no 
 *  warranties, expressed or implied.
 * 
 *  Options:
 *
 *        -acr  format actors database
 *        -acs  format actress database
 *        -dir  format directors database
 *      -write  format writers database
 *       -comp  format composers database
 *       -cine  format cinematographers database
 *       -edit  format editors database
 *     -prodes  format production designers database
 *    -costdes  format costume designers database
 *      -prdcr  format producers database
 *       -misc  format miscellaneous database
 *
 *    -mrr        add movie ratings report info to results
 *    -smrr       add movie ratings report info & sort by ratings
 *    -vmrr       add movie ratings report info & sort by votes
 *
 *    -yru        add year of release info
 *    -yr         add year of release info & sort chronologically
 *
 *    -m          movies only, ignore TV-series
 *    -r          raw, don't format results
 *
 *============================================================================
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "moviedb.h"
#include "dbutils.h"
#include "years.h"
#include "ratings.h"

#define LFORMAT_USAGE1 "usage: lformat [-acr|-acs|-dir|-write|-comp|-cine|-edit|-prodes|-costdes|"
#define LFORMAT_USAGE2 "               -prdcr|-misc] [-mrr|-smrr|-vmrr] [-yr|-yru] [-m -r]"


void displayLformatRaw ( struct formatRec listData [ ], int count ) 
{
  int i ;

  for ( i = 0 ; i < count ; i++ )
  {
    (void) printf ( "%s|", listData [ i ] . name ) ;
    if ( listData [ i ] . year != 0 )
      displayTitleYear ( listData [ i ] . title, listData [ i ] . year ) ;
    else
      (void) printf ( "%s", listData [ i ] . title ) ;
    if ( listData [ i ] . attr != NULL )
      (void) printf ( "|%s\n", listData [ i ] . attr ) ;
    else
      (void) printf ( "|\n" ) ;
  }
}


void displayLformatYear ( struct formatRec listData [], int count )
{
  int i, len ;

  (void) printf ( "%s", listData [ 0 ] . name ) ;
  len = strlen ( listData [ 0 ] . name ) ;
  if ( len >= 16 )
    (void) printf ( "\t" ) ;
  else if ( len >= 8 )
    (void) printf ( "\t\t" ) ;
  else
    (void) printf ( "\t\t\t" ) ;
    
  displayTitleYear ( listData [ 0 ] . title, listData [ 0 ] . year ) ;
  if ( listData [ 0 ] . attr != NULL )
    (void) printf ( " %s\n", listData [ 0 ] . attr ) ;
  else
    (void) printf ( "\n" ) ;

  for ( i = 1 ; i < count ; i++ )
  {
    (void) printf ( "\t\t\t" ) ;
    displayTitleYear ( listData [ i ] . title, listData [ i ] . year ) ;
    if ( listData [ i ] . attr != NULL )
      (void) printf ( " %s\n", listData [ i ] . attr ) ;
    else
      (void) printf ( "\n" ) ;
  }

  (void) printf ( "\n" ) ;
}


void displayLformatRatings ( struct formatRec listData [], int count ) 
{
  int i ;

  (void) printf ( "%s\n\n", listData [ 0 ] . name ) ;

  for ( i = 0 ; i < count ; i++ )
  {
    if ( listData [ i ] . mrr . votes > 0 )
      (void) printf ( "      %s%8d   %2.1f  ", listData [ i ] . mrr . distribution, listData [ i ] . mrr . votes, listData [ i ] . mrr . rating ) ;
    else
      (void) printf ( "                                " ) ;

    if ( listData [ i ] . year != 0 )
      displayTitleYear ( listData [ i ] . title, listData [ i ] . year ) ;
    else
      (void) printf ( "%s", listData [ i ] . title ) ;
    if ( listData [ i ] . attr != NULL )
      (void) printf ( " %s\n", listData [ i ] . attr ) ;
    else
      (void) printf ( "\n" ) ;
  }
  (void) printf ( "\n" ) ;
}


void displayLformatData ( struct formatRec listData [], int count, int raw, int mrropt, int yropt ) 
{
  if ( raw )
    displayLformatRaw ( listData, count ) ;
  else if ( mrropt == NONE && yropt == NONE )
    displayFormatPlain ( listData, count ) ;
  else if ( mrropt == NONE && yropt != NONE )
    displayLformatYear ( listData, count ) ;
  else if ( mrropt != NONE )
    displayLformatRatings ( listData, count ) ;
}


int plainAlphaSort ( struct formatRec *r1, struct formatRec *r2 )
{
  if ( r1 -> title [ 0 ] == '"' )
    if ( r2 -> title [ 0 ] == '"' )
      return ( caseCompare ( r1 -> title, r2 -> title )  ) ;
    else
      return ( 1 ) ;
  else
    if ( r2 -> title [ 0 ] == '"' )
      return ( -1 ) ;
    else
      return ( caseCompare ( r1 -> title, r2 -> title )  ) ;
}


int titleKeyLformatCompare ( struct titleDbRec *search, struct titleDbRec *t1 )
{
  return ( search -> titleKey - t1 -> titleKey ) ;
}


int mrrKeyLformatCompare ( struct mrrDbRec *search, struct mrrDbRec *m1 )
{
  return ( search -> titleKey - m1 -> titleKey ) ;
}


void sortLformatData ( struct formatRec listData [ ], int count, int raw, int mrropt, int yropt, struct titleDbRec *years, TitleID nyears, struct mrrDbRec *ratings, TitleID nratings ) 
{
  struct titleDbRec *trec, tSearchRec ;
  struct mrrDbRec *mrec, mSearchRec ;
  int i ;

  if ( mrropt == NONE && yropt == NONE && !raw )
  {
    qsort ( (void*) listData, (size_t) count, sizeof ( struct formatRec ), (int (*) (const void*, const void*)) plainAlphaSort ) ;
    return ;
  }

  if ( yropt != NONE )
  {
    for ( i = 0 ; i < count ; i++ )
    {
      tSearchRec . titleKey = listData [ i ] . titleKey ;
      if ( ( trec = bsearch ( (void*) &tSearchRec, (void*) years, (size_t) nyears, sizeof ( struct titleDbRec ), (int (*) (const void*, const void*)) titleKeyLformatCompare ) ) != NULL )
        listData [ i ] . year = trec -> year ;
      else
        listData [ i ] . year = 0 ;
    }
  }

  if ( mrropt != NONE )
  {
    for ( i = 0 ; i < count ; i++ )
    {
      mSearchRec . titleKey = listData [ i ] . titleKey ;
      if ( ( mrec = bsearch ( (void*) &mSearchRec, (void*) ratings, (size_t) nratings, sizeof ( struct mrrDbRec ), (int (*) (const void*, const void*)) mrrKeyLformatCompare ) ) != NULL )
        listData [ i ] . mrr = mrec -> mrr ;  
      else
      {
        listData [ i ] . mrr . votes = 0 ;  
        listData [ i ] . mrr . rating = 0.0 ;  
      }
    }
  }

  if ( mrropt == NONE ) 
    if ( yropt == YRU )
    {
      if ( !raw )
        qsort ( (void*) listData, (size_t) count, sizeof ( struct formatRec ), (int (*) (const void*, const void*)) plainAlphaSort ) ;
    }
    else 
      qsort ( (void*) listData, (size_t) count, sizeof ( struct formatRec ), (int (*) (const void*, const void*)) yearLformatSort ) ;

  if ( mrropt == SMRR ) 
    qsort ( (void*) listData, (size_t) count, sizeof ( struct formatRec ), (int (*) (const void*, const void*)) smrrLformatSort ) ;
  else if ( mrropt == VMRR ) 
    qsort ( (void*) listData, (size_t) count, sizeof ( struct formatRec ), (int (*) (const void*, const void*)) vmrrLformatSort ) ;
}


void lformatList ( int listId, int mvsonly, int raw, int mrropt, int yropt, struct titleDbRec *years, TitleID nyears, struct mrrDbRec *ratings, TitleID nratings )
{
  FILE *dbFp, *nameKeyFp, *nameIndexFp, *titleKeyFp, *titleIndexFp, *attrKeyFp, *attrIndexFp ;
  struct formatRec listData [ MAXFILMOGRAPHIES ] ;
  int i, count, len, noWithAttr, noWithoutAttr ;
  char fn [ MAXPATHLEN ] ;
  char *name ;
  NameID nameKey ;
  TitleID titleKey ;
  AttributeID attrKey ;

  (void) constructFilename ( fn, filmographyDefs [ listId ] .stem, DBSEXT ) ;
  dbFp = openFile ( fn ) ;
  nameKeyFp = openFile ( NAMEKEY ) ;
  nameIndexFp = openFile ( NAMEIDX ) ;
  titleKeyFp = openFile ( TITLEKEY ) ;
  titleIndexFp = openFile ( TITLEIDX ) ;
  attrKeyFp = openFile ( ATTRKEY ) ;
  attrIndexFp = openFile ( ATTRIDX ) ;

  while ( TRUE )
  {
    count = 0 ;
    nameKey = getName ( dbFp ) ;
    name = mapNameKeyToText ( nameKey, nameKeyFp, nameIndexFp ) ;
    if ( feof ( dbFp ) )
      break ;
    getFilmographyCounts ( dbFp, &noWithAttr, &noWithoutAttr ) ;
    for ( i = 0 ; i < noWithAttr ; i++ )
    {
      titleKey = getTitle ( dbFp ) ;
      attrKey = getAttr ( dbFp ) ;
      if ( listId < NO_OF_CAST_LISTS )
      {
        len = getByte ( dbFp ) ;
        if ( len > 0 )
          (void) fseek ( dbFp, len, SEEK_CUR ) ;
        (void) getPosition ( dbFp ) ;
      }
	else if ( listId == WRITER_LIST_ID )
      {
	(void) getPosition ( dbFp ) ;
	(void) getPosition ( dbFp ) ;
	(void) getPosition ( dbFp ) ;
      }
      listData [ count ] . titleKey = titleKey ;
      listData [ count ] . attrKey = attrKey ;
      listData [ count ] . title = mapTitleKeyToText ( titleKey, titleKeyFp, titleIndexFp ) ;
      if ( !mvsonly || listData [ count ] . title [ 0 ] != '"' )
      {
        listData [ count ] . name = duplicateString ( name ) ;
        listData [ count ] . attr = mapAttrKeyToText ( attrKey, attrKeyFp, attrIndexFp ) ;
        count++ ;
      }
      else
        free ( (void*) listData [ count ] . title ) ;
    }
    for ( i = 0 ; i < noWithoutAttr ; i++ )
    {
      titleKey = getTitle ( dbFp ) ;
      if ( listId < NO_OF_CAST_LISTS )
      {
        len = getByte ( dbFp ) ;
        if ( len > 0 )
          (void) fseek ( dbFp, len, SEEK_CUR ) ;
        (void) getPosition ( dbFp ) ;
      }
	else if ( listId == WRITER_LIST_ID )
      {
	(void) getPosition ( dbFp ) ;
	(void) getPosition ( dbFp ) ;
	(void) getPosition ( dbFp ) ;
      }
      listData [ count ] . attrKey = NOATTR ;
      listData [ count ] . titleKey = titleKey ;
      listData [ count ] . title = mapTitleKeyToText ( titleKey, titleKeyFp, titleIndexFp ) ;
      if ( !mvsonly || listData [ count ] . title [ 0 ] != '"' )
      {
        listData [ count ] . name = duplicateString ( name ) ;
        listData [ count ] . attr = NULL ;
        count++ ;
      }
      else
        free ( (void*) listData [ count ] . title ) ;
    }
    if ( count > 0 )
    {
      sortLformatData ( listData, count, raw, mrropt, yropt, years, nyears, ratings, nratings ) ;
      displayLformatData ( listData, count, raw, mrropt, yropt ) ;
      for ( i = 0 ; i < count ; i++ )
      {
        free ( (void*) listData [ i ] . name ) ;
        free ( (void*) listData [ i ] . title ) ;
        free ( (void*) listData [ i ] . attr ) ;
      }
    }
  }
  (void) fclose ( dbFp ) ;
  (void) fclose ( nameKeyFp ) ;
  (void) fclose ( nameIndexFp ) ;
  (void) fclose ( titleKeyFp ) ;
  (void) fclose ( titleIndexFp ) ;
  (void) fclose ( attrKeyFp ) ;
  (void) fclose ( attrIndexFp ) ;
}


int main ( int argc, char **argv )
{
  struct titleDbRec *years = NULL ;
  struct mrrDbRec   *ratings = NULL ;
  int    err = FALSE ;
  int    i, j, okopt ;
  int    mvsonly = FALSE ;
  int    raw = FALSE ;
  int    mrropt = NONE ;
  int    yropt = NONE ;
  int    listId = -1 ;
  TitleID  nyears = 0, nratings = 0 ;

  if ( argc == 1 )
    err = TRUE ;

  for ( i = 1 ; i < argc ; i++ )
    if ( strcmp ( "-m", argv [ i ] ) == 0 )
      mvsonly = TRUE ;
    else if ( strcmp ( "-r", argv [ i ] ) == 0 )
      raw = TRUE ;
    else if ( strcmp ( "-mrr", argv [ i ] ) == 0 )
      mrropt = MRR ;
    else if ( strcmp ( "-smrr", argv [ i ] ) == 0 )
      mrropt = SMRR ;
    else if ( strcmp ( "-vmrr", argv [ i ] ) == 0 )
      mrropt = VMRR ;
    else if ( strcmp ( "-yr", argv [ i ] ) == 0 )
      yropt = YR ;
    else if ( strcmp ( "-yru", argv [ i ] ) == 0 )
      yropt = YRU ;
    else
    {
      okopt = FALSE ;
      for ( j = 0 ; j < NO_OF_FILMOGRAPHY_LISTS ; j++ )
	if ( strcmp ( filmographyDefs [ j ] . option, argv [ i ] ) == 0 )
        {
	  okopt = TRUE ;
          if ( listId != -1 )
            moviedbError ( "lformat: only one database option allowed" ) ;
	  listId = j ;
	}
      if ( ! okopt )
      {
         (void) fprintf ( stderr, "lformat: unrecognised option %s\n", argv[i] ) ;
         err = TRUE ;
      }
    }

  if ( !err && listId == -1 )
  {
    (void) fprintf ( stderr, "lformat: no database specified\n" ) ;
    err = TRUE ;
  }

  if ( err )
    moviedbUsage ( LFORMAT_USAGE1, LFORMAT_USAGE2, NULL, NULL, NULL, NULL ) ;

  if ( raw && mrropt != NONE )
    moviedbError ( "lformat: -r not available with ratings report options" ) ;

  if ( yropt != NONE )
  {
    if ( ( years = calloc ( MAXTITLES, sizeof ( struct titleDbRec ) ) ) == NULL )
      moviedbError ( "out of memory" ) ;
    nyears = readTitleDb ( years ) ;
  }

  if ( mrropt != NONE )
  {
    nratings = ratingsDbSize ( ) ;
    if ( ( ratings = calloc ( nratings, sizeof ( struct mrrDbRec ) ) ) == NULL )
      moviedbError ( "out of memory" ) ;
    nratings = readRatingsDb ( ratings ) ;
  }

  lformatList ( listId, mvsonly, raw, mrropt, yropt, years, nyears, ratings, nratings ) ;

  return ( 0 ) ;
}
